src/DataSource.js

```diff
@@ -119,27 +92,92 @@ prototype.fire = function(input) {
 prototype.pipeline = function(pipeline) {
   if (!arguments.length) return this._pipeline;
 
-  var ds = this;
+  this._inputNode = DataSourceInput(this);
+  this._outputNode = DataSourceOutput(this);
+
+  var graph = this._graph,
+      mutates = 0,
+      collector = this._inputNode,
+      i, node, router, collects;
+
+  for (i=0; i<pipeline.length; ++i) {
+    node = pipeline[i];
+
+    if (!node._collector && node.batch()) {
```

Comment:
My vote would be to move this logic back to `Graph.connect`. That way vega-dataflow, as a standalone library, will correctly honor batch guarantees for all appropriately annotated Node operators, regardless of whether they're part of a DataSource or not.